### PR TITLE
Backport Service discovery hardening

### PR DESCRIPTION
**Pull request by fcrisciani**
This patch addresses several issues found on the Service Discovery feature.

### CODE REVIEW

File: endpoint.go [hunk 1]
```diff
@@ -590,6 +590,12 @@ func (ep *endpoint) rename(name string) error {
 		return fmt.Errorf("watch null for network %q", n.Name())
 	}
 
+	_, ok := ep.getSandbox()
```
Comment (fcrisciani) [inline endpoint.go]: @mavenugo a part from this extra check I did not touch the rename logic. There was a huge difference between the 2 and maybe is not worth the risk, WDYT?

File: network.go [hunk 2]
```diff
@@ -1658,15 +1661,23 @@ func (n *network) ResolveIP(ip string) string {
 
 	nwName := n.Name()
 
-	n.Lock()
-	defer n.Unlock()
-	svc, ok = sr.ipMap[ip]
-
-	if ok {
-		return svc + "." + nwName
+	nameList, ok := sr.ipMap.Get(ip)
```
Comment (mavenugo) [inline network.go]: any reason why it is named as nameList here while the master patch is called elemSet ?

Comment (fcrisciani) [inline network.go]: This is a list of strings, on master is a list of struct

**Comment by mavenugo** [comment]
LGTM

**Pull request merged**